gmt-4.5.14/src/mgg/dat2gmt.1
.TH DAT2GMT 1 "1 Nov 2015" "GMT 4.5.14" "Generic Mapping Tools"
.SH NAME
dat2gmt \- Convert an ASCII file to a binary gmt file
.SH SYNOPSIS
\fBdat2gmt\fP \fIgmtfile\fP
.SH DESCRIPTION
\fBdat2gmt\fP reads an ASCII version of a gmt file (made by \fBgmt2dat\fP) from standard
input and creates a binary gmt file.
.SH EXAMPLES
To create a gmt file from the ASCII file c2104_fixed.ascii, run
.br
.sp
\fBdat2gmt\fP c2104_fixed.ascii c2104.gmt
.SH "SEE ALSO"
.IR gmt2dat (1),
.IR GMT (1)
.SH REFERENCES
Wessel, P., and W. H. F. Smith, 2015, The Generic Mapping Tools (GMT) version
4.5.14 Technical Reference & Cookbook, SOEST/NOAA.
.br
Wessel, P., and W. H. F. Smith, 1998, New, Improved Version of Generic Mapping
Tools Released, EOS Trans., AGU, 79 (47), p. 579.
.br
Wessel, P., and W. H. F. Smith, 1995, New Version of the Generic Mapping Tools
Released, EOS Trans., AGU, 76 (33), p. 329.
.br
Wessel, P., and W. H. F. Smith, 1995, New Version of the Generic Mapping Tools
Released, http://www.agu.org/eos_elec/95154e.html, Copyright 1995 by the
American Geophysical Union.
.br
Wessel, P., and W. H. F. Smith, 1991, Free Software Helps Map and Display Data,
EOS Trans., AGU, 72 (41), p. 441.
.br


gmt-4.5.14/ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 10331 2015-10-29 19:41:56Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
2015-11-01	pwessel
	* GMT 4.5.14 finalized and released
-----------------------------------------------------------------------------------------
2015-10-29	pwessel
	* meca/psmeca.c:	Enforce rakes and strikes to be in -180/+180 range.
				Issue #799, for gmt4 only as not a problem in gmt5.

2015-07-28	pwessel
	* meca/psmeca.c:	Be able to read any value x and y coordinates

2015-06-24	pwessel
	* gmt_support.c:	Fix bug in jump near Dateline in gmt_inonout_sphpol_count,
				affecting inside/outside spherical cases, reported in message-2219

2015-03-18	pwessel
	* share/cpt/GMT_wysiwyg.cpt:	Fix typo in blue for high end [Issue #689].

2015-01-16	pwessel
	* gmt_plot.c:		Fix issue # 662 with graph arrows for negative scales.

2015-01-10	pwessel
	* gmt_plot.c:		Flipping the n/s sides array when lon and -y axes
				fixes issue #520

2015-01-01	pwessel
	* configure.ac:		Moved to 4.5.14 and back to svn release only.
-----------------------------------------------------------------------------------------
2015-01-01	pwessel
	* GMT 4.5.13 finalized and released [GSHHG 2.3.4 was released on Jan 1 as well]
-----------------------------------------------------------------------------------------
2014-12-28	pwessel
	* gmt_map.c:		Needed separate GMT_cartesian_overlap for non-geographic
				projections.  Fixes issue # 657.

2014-09-29	pwessel
	* gmt_plot.c:		Left a PSL variable undefined when grdimage plotted a grid
				with x = longitude and y = Cartesian.

2014-09-09	pwessel
	* grdfilter.c:		Fix y-shifts when -I sets smaller output.  Issue #616.
	* gmt_bcr.c:		Fix problem with points slightly outside causing an
				access violation; see issue #620.

2014-08-29	pwessel
	* sample1d.c:		Fix resampling for decreasing t values.  Issue #608.

2014-08-05	pwessel
	* gmt_map.c:		Could run into trouble clipping polygons for non-periodic
				maps.  Added more checks in GMT_wesn_clip.

2014-07-31	pwessel
	* gmt_map.c:		Rect clipping with -R<w/s/e/n>r could introduce stray lines
				when dataset has features about 180 degrees away from
				the user's area. See issue # 594.

2014-07-30	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.2 (released August 1, 2014)
	* psxy[z].c:		Fixed wrap-around issue in fault lines.

2014-07-27	pwessel
	* gmtselect.c:		The -Z option needed to pass records with NaNs.

2014-07-19	pwessel
	* x2sys/x2sys_solve.c:	Fix problems that arise when a track has less crossovers
				than model parameters.
	* gmtmath.c:		Fix issue #123 with files and -C.

2014-07-01	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.1 (released July 1, 2014)
	
2014-06-28	pwessel
	* mgd77/mgd77sniffer.c:	Needed to check that grid values used in decimate function
	 			also were within reasonable range.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je/E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
2014-05-25	pwessel
	* grdmath.c:		Fix issue #566 bug in grdmath's KM2DEG operator.

2014-03-19	pwessel
	* gmtselect.c:		The -Fpolygon option might miss points because the io-machinery
				would sometimes set the min/max longitudes found incorrectly.

2014-03-14	pwessel
	* gmt_shore.c:		Give more error feedback if nc_open fails due to old library.

2014-03-01	pwessel
	* configure.ac:		Moved to 4.5.13 and back to svn release only.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	* GMT 4.5.12 finalized and released [GSHHG 2.3.0 was released on Feb 1]
-----------------------------------------------------------------------------------------
2014-02-17	fwobbe
	* gmt_init.c etc.:	File locking -- if enabled via -DFLOCK -- is
				non-compulsory now.  This means if fcntl fails for some reason,
				GMT continues to access the file.  Hence, it is recommended to
				configure GMT with --enable-flock even if fcntl might be buggy
				on some ancient NFS shares.

2014-02-17	pwessel
	* gmt_map.c:		Size for memory allocation cast the product of nx * ny to
				size_t but overflow had already happen as calculation was
				done using ints, then cast to size_t. Must cast nx and ny
				separately.
2014-02-10	jluis
	* misc/gmt2kml.c:	Make sure the default style IDs are unique for each process.

2014-02-07	jluis
	* ps2raster.c:		Make sure that images in KML files do not carry a directory name

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-02-01	pwessel
	* configure.ac:		Since GSHHG 2.3.0 was released today, make that the default

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.

2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		Had to let SDIST return km instead of degrees since it cannot
				do degrees for geodesics.  By doing km it is now aligned with
				what PDIST and LDIST returns for geographic data, and matches
				what GMT5 does.  Some scripts had to be modified to handle the
				new behavior.  If you use SDIST you may wish to check your usage.
				Also fixed bug in SDIST in that spherical selected geodesics

gmt-4.5.14/src/mgg/dat2gmt.c
/*--------------------------------------------------------------------
 *	$Id: dat2gmt.c 10289 2014-12-28 21:17:06Z pwessel $
 *
 *    Copyright (c) 1991-2015 by P. Wessel and W. H. F. Smith
 *    See README file for copying and redistribution conditions.
 *--------------------------------------------------------------------*/
/*
 * dat2gmt will read an *.dat file (created by gmt2dat), decode it
 * and create a new *.gmt file. To edit a *.gmt file, run gmt2dat
 * to make an ASCII listing, use the editor to change the contents,
 * and then run dat2gmt to create a new binary *.gmtfile.
 *
 * Author:	Paul Wessel
 * Date:	8-FEB-1988
 * Version:	2.0
 * Version:	3.2 12-APR-1999 (DOS compatible)
 *
 */

#include "gmt.h"
#include "gmt_mgg.h"

#define MEM_SIZE 65536

int main (int argc, char **argv)
{
	int leg_year, n_records, rec, time, mag, top, yr, mo, dy, hr, mi, sc, n_alloc = MEM_SIZE, conv;
	double lat, lon, grv;
	char line[BUFSIZ], agency[10], stime[20], sgrv[10], smag[10], stop[10], *not_used = NULL;
	FILE *fpi = NULL, *fpo = NULL;
	struct GMTMGG_TIME *gmt;
	struct GMTMGG_REC *record = NULL;

	argc = (int)GMT_begin (argc, argv);
	
	if (argc != 3 || ((fpi = fopen(argv[1], "r")) == NULL) || ((fpo = fopen(argv[2], "wb")) == NULL)) {
		fprintf(stderr, "usage: dat2gmt asciifile file.gmt\n");
		exit (EXIT_FAILURE);
	}
	
	record = (struct GMTMGG_REC *) GMT_memory (VNULL, (size_t)n_alloc, sizeof (struct GMTMGG_REC), "dat2gmt");
	
	/* Read first record of file containing start-year, n_records and info */
	
	not_used = fgets (line, BUFSIZ, fpi);
	sscanf (line, "%d %d %s", &leg_year, &n_records, agency);
	gmt = gmtmgg_init (leg_year);
	
	rec = 0;
	while (fgets (line, BUFSIZ, fpi)) {
		sscanf (line, "%s %lf %lf %s %s %s", stime, &lat, &lon, sgrv, smag, stop);
		sscanf (stime, "%d/%d/%d/%d:%d:%d", &yr, &mo, &dy, &hr, &mi, &sc);
		record[rec].time = gmtmgg_time (&time, yr, mo, dy, hr, mi, sc, gmt);
		record[rec].lat = (int) (lat * 1000000.0);
		record[rec].lon = (int) (lon * 1000000.0);
		conv = sscanf(sgrv, "%lf", &grv);
		record[rec].gmt[0] = (conv > 0 && !GMT_is_dnan (grv)) ? (int) rint(grv * 10.0) : GMTMGG_NODATA;
		conv = sscanf(smag, "%d", &mag);
		record[rec].gmt[1] = (conv > 0) ? mag : GMTMGG_NODATA;
		conv = sscanf(stop, "%d", &top);
		record[rec].gmt[2] = (conv > 0) ? top : GMTMGG_NODATA;
		rec++;
		if (rec == n_alloc) {
			n_alloc <<= 1;
			record = (struct GMTMGG_REC *) GMT_memory((void *)record, (size_t)n_alloc, sizeof (struct GMTMGG_REC), "dat2gmt");
		}
	}
	fclose (fpi);
	
	GMT_free ((void *)gmt);
	n_records = rec;
	if (fwrite((void *)(&leg_year), (size_t)4, (size_t)1, fpo) != 1) {
		fprintf(stderr,"dat2gmt: Error while writing first year\n");
		exit (EXIT_FAILURE);
	}
	if (fwrite((void *)(&n_records), (size_t)4, (size_t)1, fpo) != 1) {
		fprintf(stderr,"dat2gmt: Error while writing no of records\n");
		exit (EXIT_FAILURE);
	}
	if (fwrite((void *)agency, (size_t)10, (size_t)1, fpo) != 1) {
		fprintf(stderr,"dat2gmt: Error while writing info-header\n");
		exit (EXIT_FAILURE);
	}
	for (rec = 0; rec < n_records; rec++) {
		if (fwrite ((void *)(&record[rec]), (size_t)18, (size_t)1, fpo) != 1) {
			fprintf (stderr,"dat2gmt: Error writing data record no %d\n",rec);
			exit (EXIT_FAILURE);
		}
	}
	fclose (fpo);
	
	GMT_free ((void *)record);

	GMT_end (argc, argv);
	
	exit (EXIT_SUCCESS);
}
